src/index.ts

```diff
@@ -53,6 +54,45 @@ export default class FiatConnectClient implements FiatConectApiClient {
     }
   }
 
+  /**
+   * https://en.wikipedia.org/wiki/Network_Time_Protocol#Clock_synchronization_algorithm
+   *
+   * Convenience method to calculate the difference between server and client clocks.
+   * Returns the difference between the client and server clocks as a number of milliseconds.
+   * Positive values mean the server's clock is ahead of the client's.
+   */
+  async getClockDiff(): Promise<Result<number, ErrorResponse>> {
```

Comment:
I suggest we rename this to getClockDiffApprox, and return the max error `(t3-t0)/2` in addition to the estimated diff. Or just return the min and max possible values for theta, which are `t1-t3` and `t1-t0`, and don't make any guesses about network variation, which could be safer.

Even assuming `t1=t2`, which seems fine, we don't have enough information to get the *actual* clock diff because we don't know how much of the network delay is in the "to" trip vs the "return" trip. 

Say the "to" trip delay is `d1` and the "from" trip delay is `d2`. We have three equations and three unknowns, so at first it looks fine:
```
d1 + d2 = t3 - t0
t0 + d1 + theta = t1
t3 + theta - d2 = t2
```
However, if you subtract the second equation from the third, you get:
```
t3 - t0 - d1 - d2 = t2 - t1
```
So with our assumption that `t2 = t1`, we get:
```
t3 - t0 - d1 - d2 = 0
```
And we're hosed! It's the same as our first equation; we cannot solve for theta. (The equation matrix has a nontrivial null space 🙂 .)

The algo quietly assumes that `d1=d2`, which gives enough to solve again. However, that doesn't seem like a guarantee by any means. 

The reason I say all this is because if our approximation is wrong-- which it will be, almost all the time, by some amount-- that could mess us up, for example if we use this to set an `issued-at` timestamp on a login message that the server then checks.

Reply:
Thanks for the detailed write up. I definitely realize that this is only a rough approximation. I like the idea of also returning the max error along with the calculated diff; I've updated the implementation to support this.

Reply:
Just for my understanding, how is `maxError = (t3 - t0) / 2`?
Also, since this is approx, how would clients use this?

Reply:
The implicit assumption in the above calculation is that the "to" trip and "from" trips are the same duration. (NTP actually makes this assumption; it could be the case that statistically, this holds when averaged over many requests, but we're just doing one here so it's kind of moot.) Intuitively, we assume that the time reported by the server is their clock at the moment halfway between `t0` and `t3`, when in the "worst case", it could _actually_ be immediately after `t0` or immediately before `t3`. In either case, the max error introduced from our assumption of symmetry in trip times would be `(t3 - t0) / 2`, the distance between the midpoint of `t0` and `t3` (`(t0 + t3) /2 `), and `t0` or `t3`.

As for how clients would use this, probably for worst case estimations. Since auth will fail if the `issued-at` time is before the server's timestamp, clients can use the max error to skew their estimation towards the server's clock being later.